Changes for docker-based apps

mattdrees opened the pull request:
This PR contains a few changes that make this client usable by a docker-based app.

This solves two problems:
1. Docker apps typically use the same image in various environments, and use environment variables to drive environment-specific config. The client's current web.xml-based configuration makes this impossible.
2. Docker apps typically do not run with statically-assigned hostnames. So, the LogoutEchoFilter cannot be used, as it requires a static list of hostnames.

We solve the first by allowing jndi-based config (which, in wildfly, can be driven by environment variables), and the second by using a distributed cache to store logout requests instead of a static in-vm list.

mattdrees committed:
allow parameters to be driven by jndi

This is roughly how the more modern cas 3.1 client works:
https://wiki.jasig.org/display/CASC/Configuring+the+JA-SIG+CAS+Client+for+Java+using+JNDI

Hopefully, if we are able to switch to the cas 3.1 client, this will reduce the configuration changes necessary.

mattdrees committed:
add convenience parameter 'casServerUrlPrefix'

This drives both the loginUrl and the validateUrl parameters, making configuration simpler.
It's also consistent with the cas 3.1 client configuration (for the saml validator, at least).

mattdrees committed:
bump version to 2.1.6

mattdrees commented:
@hlbraddock can you review for me?

hlbraddock commented:
Could you not switch to CAS client 3.1 in order to address problem #1?

hlbraddock commented:
I'm a little fuzzy on how docker images work with varying configurations. I suppose I assumed that all configuration was part of that docker image (the docker image being 'everything'). How does a docker image use environment variables not a part of that image?

mattdrees commented:
The official CAS client 3.1? I'm not sure of a good way to do that, since we need access to custom attributes in the cas payload, and we need to support single logout, and I don't think the official client supports either of these  (or at least, it doesn't in the way that we use them. I think the saml version supports these, but that's something I didn't want to have to switch to that yet.)

hlbraddock commented on src/main/java/edu/yale/its/tp/cas/client/filter/CASFilter.java:
```diff
@@ -289,17 +296,35 @@
+        casServerName = Configuration.getParameter(config, SERVERNAME_INIT_PARAM);
+        casProxyCallbackUrl = Configuration.getParameter(config, PROXY_CALLBACK_INIT_PARAM);
+        casLogoutCallbackUrl = Configuration.getParameter(config, LOGOUT_CALLBACK_INIT_PARAM);
+        wrapRequest = Boolean.valueOf(Configuration.getParameter(config, WRAP_REQUESTS_INIT_PARAM)).booleanValue();
```
Boolean value unboxing is unnecessary (Java 5 and after).

`Boolean.valueOf(Configuration.getParameter(config, <param name>)).booleanValue();`

can be simplified to:

`Boolean.valueOf(Configuration.getParameter(config, <param name>));`

mattdrees commented:
When you start up a docker container, you can feed it additional environment variables. Some env variables can be part of the image, but not the ones that need to vary per-deployment.

hlbraddock commented on src/main/java/edu/yale/its/tp/cas/client/filter/CASFilter.java:
```diff
@@ -289,17 +296,35 @@
+        wrapRequest = Boolean.valueOf(Configuration.getParameter(config, WRAP_REQUESTS_INIT_PARAM)).booleanValue();
+        casGateway = Boolean.valueOf(Configuration.getParameter(config, GATEWAY_INIT_PARAM)).booleanValue();
+        remoteUserAttrib = Configuration.getParameter(config, REMOTE_USER_ATTRIB_INIT_PARAM);
+
```
Not part of this commit, but also: `Boolean.valueOf(casRenew).booleanValue())` can be simplified to `casRenew`.

mattdrees replied:
I've removed all the unnecessary Boolean unboxing from the project

hlbraddock commented:
That's true. I suppose it's a lot more work to customize CAS 3.1 than to make the changes you have made here.

hlbraddock commented:
OK, so feeding environment variables to a docker image (at deploy time, I presume) is standard procedure, in which case it seems JNDI environment variable lookup is a logical replacement (for web.xml configuration).

mattdrees committed:
remove unnecessary unboxing

(and in two cases, boxing and unboxing)

mattdrees commented:
infinispan is what is already present in wildfly

mattdrees commented:
sorry, missed your second question. The library is a module in the wildfly distribution. Wildfly needs some config to set up the infinispan cache; that config is located here: https://github.com/CruGlobal/ecs_config/blob/master/ecs/base-image-wildfly/cas.cli

So our apps (that use base-image-wildfly) only need to make sure cas.cli gets called, like this:
https://github.com/CruGlobal/budget-tool/blob/dockerize/src/main/docker/Dockerfile#L10

hlbraddock commented on src/main/java/edu/yale/its/tp/cas/util/LogoutEchoFilter.java:
```diff
@@ -31,7 +32,10 @@
- * 
+ *
+ * This is useful if the host names of the other instances can be determined up-front.
+ * It is not needed when using {@link InfinispanLogoutStorage}.
```
+1 Nice Javadoc link.

hlbraddock commented on src/main/java/edu/yale/its/tp/cas/client/filter/CASFilter.java:
```diff
@@ -22,6 +22,7 @@
 import javax.servlet.http.HttpServletResponse;
 import javax.servlet.http.HttpSession;
 
+import edu.yale.its.tp.cas.util.Configuration;
```
I don't quite see how `https://github.com/CruGlobal/cas-client/blob/changes-for-docker/src/main/java/edu/yale/its/tp/cas/client/filter/CASFilter.java#L474` will ever be true for the 'other' cas client nodes (those who didn't receive the actual logout request, but rather are relying on shared cache), since their won't (or might not) be a 'logout' (preceding -) service ticket in the request.

mattdrees replied:
so `isReceiptQueuedForLogout()` checks for the service ticket from the `CASReceipt` (from the session), not from the http request.  So it should work. Perhaps, though, I'm misunderstanding your question?

hlbraddock replied:
Nope, you are correct and understood my question. I mistakenly assumed CASReceipt was concocted from the http request.

hlbraddock commented:
I suppose you don't deprecate `LogoutEchoFilter` for those multi server CAS clients who prefer not to use a shared Infinispan distributed cache? Otherwise, I think it's the preferable solution. Echoing logout seemed a bit of a rigmarole to have to go through.

hlbraddock commented:
Side note: https://github.com/CruGlobal/cas-client/blob/changes-for-docker/src/main/java/edu/yale/its/tp/cas/client/filter/CASFilter.java#L550 - 563 is always false and appears to be duplicate from code block https://github.com/CruGlobal/cas-client/blob/changes-for-docker/src/main/java/edu/yale/its/tp/cas/client/filter/CASFilter.java#L487 - 500.

hlbraddock commented:
Is https://github.com/CruGlobal/ecs_config/blob/master/ecs/base-image-wildfly/cas.cli#L3 what makes `CasServerUrlPrefix` environment variable available for JNDI lookup, and if so, where are the other `Configuration` environment variables made available for JNDI lookup?

mattdrees commented:
Wow, that is bizarre. I wonder how that came about. Looks like it happened in baaac09026796bae6bc3ddacf16d49a20ac0f5b6.

I think we can safely remove the second block, then, right?

mattdrees commented:
Close, that line is what makes the `CAS_SERVER_URL_PREFIX` environment variable map to the `CasServerUrlPrefix` JNDI location.

It turns out this is the only JNDI config option I needed to set, at least for the budget tool. If other apps need other options, we'll have to tweak cas.cli (or the app can add its own .cli script to add them).

hlbraddock commented:
+1 Looks good.

hlbraddock commented:
Yes, I think you can safely remove the second block.

mattdrees commented:
Thanks Lee. I'll split up that commit, and then merge this.

mattdrees committed:
remove dead code

This was probably accidentally added in baaac09026796bae6bc3ddacf16d49a20ac0f5b6

hlbraddock commented:
I suppose I don't quite understand https://github.com/CruGlobal/ecs_config/blob/master/ecs/base-image-wildfly/cas.cli#L12 as appropriate syntax to make infinispan cache available for JNDI lookup.  https://github.com/CruGlobal/ecs_config/blob/master/ecs/base-image-wildfly/cas.cli#L3 makes much more sense to me.

mattdrees commented:
So, the second is simply putting a String in JNDI. It's pretty straightforward.

The first is a little different, since we're telling wildfly to create an infinispan cache. We don't tell it to put this in jndi, because it does it by default, in `java:jboss/infinispan/container/{name}`. 

The wildfly docs don't seem to mention, this however. Perhaps I should put a comment to this affect on cas.cli#L12.

hlbraddock commented:
OK. That does make more sense.

mattdrees committed:
refactor JNDI utility to handle more than just Strings

This renames it to 'Configuration', since this is not just for looking up string parameters anymore.

mattdrees committed:
add ability to store 'logout list' in infinispan

This abstracts the storage of logout requests, and makes it possible to store them in a distributed cache (infinispan) instead of a static in-jvm list.
With this, the LogoutEchoFilter is not needed. This allows docker-based apps to correctly handle logout requests, because each node will not have a static hostname.

mattdrees commented:
Ok, that commit is now split in two. (Previous head was 2b9ff07).
